### PR TITLE
Closes #221

### LINKED ISSUE #221: Create dataset loader for n2c2 2018 - Clinical Trial Cohort Selection
## Adding a Dataset
- **Name:** n2c2 2018 - Clinical Trial Cohort Selection
- **Description:** *None provided*
- **Task:** DOC_CLASS
- **Paper:** https://academic.oup.com/jamia/article-abstract/26/11/1163/5575392?redirectedFrom=fulltext
- **Data:** https://portal.dbmi.hms.harvard.edu/projects/n2c2-nlp/
- **License:** DUA-C/NC

### PR DESCRIPTION
**Pull request by singhay**
Please name your PR after the issue it closes. You can use the following line: "Closes #ISSUE-NUMBER" where you replace the ISSUE-NUMBER with the one corresponding to your dataset.

If the following information is NOT present in the issue, please populate:

- **Name:** *name of the dataset*
- **Description:** *short description of the dataset (or link to social media or blog post)*
- **Paper:** *link to the dataset paper if available*
- **Data:** *link to the online home of the dataset*

### Checkbox

- [x] Confirm that this PR is linked to the dataset issue.
- [x] Create the dataloader script `biodatasets/my_dataset/my_dataset.py` (please use only lowercase and underscore for dataset naming).
- [x] Provide values for the `_CITATION`, `_DATASETNAME`, `_DESCRIPTION`, `_HOMEPAGE`, `_LICENSE`, `_URLs`, `_SUPPORTED_TASKS`, `_SOURCE_VERSION`, and `_BIGBIO_VERSION` variables.
- [x] Implement `_info()`, `_split_generators()` and `_generate_examples()` in dataloader script.
- [x] Make sure that the `BUILDER_CONFIGS` class attribute is a list with at least one `BigBioConfig` for the source schema and one for a bigbio schema.
- [x] Confirm dataloader script works with `datasets.load_dataset` function.
- [x] Confirm that your dataloader script passes the test suite run with `python -m tests.test_bigbio biodatasets/my_dataset/my_dataset.py`.
- [x] If my dataset is local, I have provided an output of the unit-tests in the PR (please copy paste). This is OPTIONAL for public datasets, as we can test these without access to the data files.

```
❯ python -m tests.test_bigbio biodatasets/n2c2_2018/n2c2_2018_track1.py --data_dir /Users/ayush.singh/workspace/data/Healthcare/2018_Clinical_Trial_Cohort_Selection
INFO:__main__:args: Namespace(path='biodatasets/n2c2_2018/n2c2_2018_track1.py', schema=None, subset_id=None, data_dir='/Users/ayush.singh/workspace/data/Healthcare/2018_Clinical_Trial_Cohort_Selection', use_auth_token=None)
INFO:__main__:self.PATH: biodatasets/n2c2_2018/n2c2_2018_track1.py
INFO:__main__:self.SUBSET_ID: n2c2_2018_track1
INFO:__main__:self.SCHEMA: None
INFO:__main__:self.DATA_DIR: /Users/ayush.singh/workspace/data/Healthcare/2018_Clinical_Trial_Cohort_Selection
INFO:__main__:Checking for _SUPPORTED_TASKS ...
INFO:__main__:Found _SUPPORTED_TASKS=[<Tasks.TEXT_CLASSIFICATION: 'TXTCLASS'>]
INFO:__main__:_SUPPORTED_TASKS implies _MAPPED_SCHEMAS={'TEXT'}
INFO:__main__:schemas_to_check: {'TEXT'}
INFO:__main__:Checking load_dataset with config name n2c2_2018_track1_source
WARNING:datasets.builder:Using custom data configuration n2c2_2018_track1_source-d42ee214038625b8
Downloading and preparing dataset n2_c22018_cohort_selection_dataset/n2c2_2018_track1_source to /Users/ayush.singh/.cache/huggingface/datasets/n2_c22018_cohort_selection_dataset/n2c2_2018_track1_source-d42ee214038625b8/1.0.0/b264439d659d3323e7bba128ca52234c37092405ae44fb9d4b1712012445fae8...
Dataset n2_c22018_cohort_selection_dataset downloaded and prepared to /Users/ayush.singh/.cache/huggingface/datasets/n2_c22018_cohort_selection_dataset/n2c2_2018_track1_source-d42ee214038625b8/1.0.0/b264439d659d3323e7bba128ca52234c37092405ae44fb9d4b1712012445fae8. Subsequent calls will reuse this data.
100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 2/2 [00:00<00:00, 126.75it/s]
INFO:__main__:Checking load_dataset with config name n2c2_2018_track1_bigbio_text
WARNING:datasets.builder:Using custom data configuration n2c2_2018_track1_bigbio_text-d42ee214038625b8
Downloading and preparing dataset n2_c22018_cohort_selection_dataset/n2c2_2018_track1_bigbio_text to /Users/ayush.singh/.cache/huggingface/datasets/n2_c22018_cohort_selection_dataset/n2c2_2018_track1_bigbio_text-d42ee214038625b8/1.0.0/b264439d659d3323e7bba128ca52234c37092405ae44fb9d4b1712012445fae8...
Dataset n2_c22018_cohort_selection_dataset downloaded and prepared to /Users/ayush.singh/.cache/huggingface/datasets/n2_c22018_cohort_selection_dataset/n2c2_2018_track1_bigbio_text-d42ee214038625b8/1.0.0/b264439d659d3323e7bba128ca52234c37092405ae44fb9d4b1712012445fae8. Subsequent calls will reuse this data.
100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 2/2 [00:00<00:00, 209.73it/s]
INFO:__main__:Checking global ID uniqueness
INFO:__main__:Found 86 unique IDs
INFO:__main__:Gathering schema statistics
INFO:__main__:Gathering schema statistics
train
==========
id: 202
document_id: 202
text: 202
labels: 2626

test
==========
id: 86
document_id: 86
text: 86
labels: 1118

.
----------------------------------------------------------------------
Ran 1 test in 0.893s

OK

```

### CODE REVIEW

**Comment by singhay** [comment]
Closes #221

File: biodatasets/n2c2_2018_track1/n2c2_2018_track1.py [hunk 1]
```diff
@@ -0,0 +1,296 @@
...
+    BIGBIO_VERSION = datasets.Version(_BIGBIO_VERSION)
+
+    # You will be able to load the "source" or "bigbio" configurations with
+    # ds_source = datasets.load_dataset('my_dataset', name='source')
+    # ds_bigbio = datasets.load_dataset('my_dataset', name='bigbio')
+
+    # For local datasets you can make use of the `data_dir` and `data_files` kwargs
+    # https://huggingface.co/docs/datasets/add_dataset.html#downloading-data-files-and-organizing-splits
+    # ds_source = datasets.load_dataset('my_dataset', name='source', data_dir="/path/to/data/files")
+    # ds_bigbio = datasets.load_dataset('my_dataset', name='bigbio', data_dir="/path/to/data/files")
```
Comment (galtay) [inline biodatasets/n2c2_2018_track1/n2c2_2018_track1.py:173]: same comments about comment removal as in #469 https://github.com/bigscience-workshop/biomedical/pull/469/files#r854751488

File: biodatasets/n2c2_2018_track1/n2c2_2018_track1.py [hunk 2]
```diff
@@ -0,0 +1,296 @@
...
+                gen_kwargs={
+                    "file_path": os.path.join(data_dir, "train.zip"),
+                    "split": str(datasets.Split.TRAIN),
+                },
+            ),
+            datasets.SplitGenerator(
+                name=datasets.Split.TEST,
+                gen_kwargs={
+                    "file_path": os.path.join(data_dir, "n2c2-t1_gold_standard_test_data.zip"),
+                    "split": str(datasets.Split.TEST),
```
Comment (galtay) [inline biodatasets/n2c2_2018_track1/n2c2_2018_track1.py:259]: these kwargs need to be consistent with what `_generate_examples` accepts.

**Review by galtay — CHANGES_REQUESTED** [review]
remove comments and `_generate_examples` kwargs

File: biodatasets/n2c2_2018_track1/n2c2_2018_track1.py (hunk 1 above)
Comment (singhay) [reply biodatasets/n2c2_2018_track1/n2c2_2018_track1.py]: I assumed these would be of help to the end user, removed in recent commit. :)

File: biodatasets/n2c2_2018_track1/n2c2_2018_track1.py (hunk 2 above)
Comment (singhay) [reply biodatasets/n2c2_2018_track1/n2c2_2018_track1.py]: done

File: biodatasets/n2c2_2018_track1/n2c2_2018_track1.py [hunk 3]
```diff
@@ -0,0 +1,296 @@
+# coding=utf-8
+# Copyright 2022 The HuggingFace Datasets Authors and
+#
+# * Ayush Singh (singhay)
```
[the remark this answers is not in the document]
Comment (singhay) [reply biodatasets/n2c2_2018_track1/n2c2_2018_track1.py]: done

File: biodatasets/n2c2_2018_track1/n2c2_2018_track1.py [hunk 4]
```diff
@@ -0,0 +1,282 @@
...
+    d. Ischemia, past or present
+8. MI-6MOS: MI in the past 6 months
+9. KETO-1YR: Diagnosis of ketoacidosis in the past year
+10. DIETSUPP-2MOS: Taken a dietary supplement (excluding vitamin D) in the past 2 months
+11. ASP-FOR-MI: Use of aspirin to prevent MI
+12. HBA1C: Any hemoglobin A1c (HbA1c) value between 6.5% and 9.5%
+13. CREATININE: Serum creatinine > upper limit of normal
+
+The training consists of 202 patient records with document-level annotations, 10 records
+with textual spans indicating annotator’s evidence for their annotations while test set contains 86.
```
Comment (galtay) [inline biodatasets/n2c2_2018_track1/n2c2_2018_track1.py:115]: are the 10 records with span level annotations left out?

File: biodatasets/n2c2_2018_track1/n2c2_2018_track1.py [hunk 5]
```diff
@@ -0,0 +1,282 @@
...
+        "MAKES-DECISIONS",
+        "MI-6MOS",
+    ]
+
+    def _info(self) -> datasets.DatasetInfo:
+
+        if self.config.schema == SOURCE:
+            labels = {
+                key: datasets.ClassLabel(names=["met", "not met"])
+                for key in N2C22018CohortSelectionDataset.LABEL_CLASS_NAMES
```
Comment (galtay) [inline biodatasets/n2c2_2018_track1/n2c2_2018_track1.py:206]:
```suggestion
                for key in self.LABEL_CLASS_NAMES
```

File: biodatasets/n2c2_2018_track1/n2c2_2018_track1.py [hunk 6]
```diff
@@ -0,0 +1,282 @@
...
+            "text": sample.get("txt", ""),
+            "tags": sample.get("tags", {}),
+        }
+
+    @staticmethod
+    def _get_bigbio_sample(sample_id, sample) -> dict:
+
+        tags = sample.get("tags", None)
+        if tags:
+            labels = [name + "-" + met_status for name, met_status in tags.items()]
```
Comment (galtay) [inline biodatasets/n2c2_2018_track1/n2c2_2018_track1.py:264]: I think its cleaner if we just include the labels that are ==`met` here ... not an entry for each with `-met` or `-not met` appended. e.g. not this 
```
['ABDOMINAL-not met',
  'ADVANCED-CAD-met',
  'ALCOHOL-ABUSE-not met',
  'ASP-FOR-MI-met',
  'CREATININE-not met',
  'DIETSUPP-2MOS-met',
  'DRUG-ABUSE-not met',
  'ENGLISH-met',
  'HBA1C-not met',
  'KETO-1YR-not met',
  'MAJOR-DIABETES-met',
  'MAKES-DECISIONS-met',
  'MI-6MOS-met']
```
but this 
```
[ 'ADVANCED-CAD',
  'ASP-FOR-MI',
  'DIETSUPP-2MOS',
  'ENGLISH',
  'MAJOR-DIABETES',
  'MAKES-DECISIONS',
  'MI-6MOS']
```

File: biodatasets/n2c2_2018_track1/n2c2_2018_track1.py [hunk 7]
```diff
@@ -0,0 +1,282 @@
...
+
+        return {"id": sample_id, "document_id": sample_id, "text": sample.get("txt", ""), "labels": labels}
+
+    def _generate_examples(self, file_path) -> (int, dict):
+        samples = _read_zip(file_path)
+
+        _id = 0
+        for sample_id, sample in samples.items():
+
+            if self.config.name == N2C22018CohortSelectionDataset._SOURCE_CONFIG_NAME:
```
Comment (galtay) [inline biodatasets/n2c2_2018_track1/n2c2_2018_track1.py:276]:
```suggestion
            if self.config.name == self._SOURCE_CONFIG_NAME:
```

File: biodatasets/n2c2_2018_track1/n2c2_2018_track1.py [hunk 8]
```diff
@@ -0,0 +1,282 @@
...
+                yield _id, self._get_source_sample(sample_id, sample)
+            elif self.config.name == N2C22018CohortSelectionDataset._BIGBIO_CONFIG_NAME:
```
Comment (galtay) [inline biodatasets/n2c2_2018_track1/n2c2_2018_track1.py:278]:
```suggestion
            elif self.config.name == self._BIGBIO_CONFIG_NAME:
```

File: biodatasets/n2c2_2018_track1/n2c2_2018_track1.py (hunk 4 above)
Comment (singhay) [reply biodatasets/n2c2_2018_track1/n2c2_2018_track1.py:115]: I had that doubt as well. However, I could not find any information on the whereabouts of those 10 records. Furthermore, I did a simple script based validation where I checked if any of the `tags` contained any text in any of the training set and they did not which confirms that atleast train and test do not have any evidence tagged alongside corresponding tags.

```python
for file in os.listdir("./train"):
    e = etree.parse("./train/"+file)

    root = e.getroot()
    text, tags = root.getchildren()
    for child in tags:
        if child.text:
            print(child.tag, child.get("met"), child.text)
```

File: biodatasets/n2c2_2018_track1/n2c2_2018_track1.py (hunk 6 above)
Comment (singhay) [reply biodatasets/n2c2_2018_track1/n2c2_2018_track1.py]: I had similar thoughts but wasn't sure about how the bigBio wanted it, better to be more explicit than loosing the information.

* By doing this we're making conscious decision of treating met == presence of `tag` and absence as literally the tag being absent.
* A downside of this approach would be for anyone to find out total unique labels they'd have to iterate all over the dataset and create a set while existing approach simply states presence/absence.
* By having arrays of variable length, it might make it harder to e.g. do one-hot encoding.

I've updated nevertheless as per your suggestion

File: biodatasets/n2c2_2018_track1/n2c2_2018_track1.py (hunk 5 above)
Comment (singhay) [reply biodatasets/n2c2_2018_track1/n2c2_2018_track1.py]: done

File: biodatasets/n2c2_2018_track1/n2c2_2018_track1.py (hunk 7 above)
Comment (singhay) [reply biodatasets/n2c2_2018_track1/n2c2_2018_track1.py]: done

File: biodatasets/n2c2_2018_track1/n2c2_2018_track1.py (hunk 8 above)
Comment (singhay) [reply biodatasets/n2c2_2018_track1/n2c2_2018_track1.py]: done

File: biodatasets/n2c2_2018_track1/n2c2_2018_track1.py (hunk 4 above)
Comment (galtay) [reply biodatasets/n2c2_2018_track1/n2c2_2018_track1.py:115]: OK, lets leave a note about this in the description then.

File: biodatasets/n2c2_2018_track1/n2c2_2018_track1.py [hunk 9]
```diff
@@ -0,0 +1,282 @@
...
+_BIGBIO_VERSION = "1.0.0"
+
+# Constants
+SOURCE = "source"
+BIGBIO_TEXT = "bigbio_text"
+
+
+def _read_zip(file_path, samples=None):
+    if samples is None:
+        samples = defaultdict(dict)
```
Comment (galtay) [inline biodatasets/n2c2_2018_track1/n2c2_2018_track1.py:137]: if `samples` is never passed in, we don't need this logic, we can simply set
`samples = defaultdict(dict)` at the start.

File: biodatasets/n2c2_2018_track1/n2c2_2018_track1.py (hunk 4 above)
Comment (singhay) [reply biodatasets/n2c2_2018_track1/n2c2_2018_track1.py:115]: done

File: biodatasets/n2c2_2018_track1/n2c2_2018_track1.py (hunk 9 above)
Comment (singhay) [reply biodatasets/n2c2_2018_track1/n2c2_2018_track1.py]: true, sorry this method I directly took from your 2011 script. Updated!

Comment (galtay) [reply biodatasets/n2c2_2018_track1/n2c2_2018_track1.py]: I recognized it :) in the 2011 script I had to build the samples from two files so needed to have the option to pass in existing samples, but here I think it would just confuse future reviewers.

Comment (singhay) [reply biodatasets/n2c2_2018_track1/n2c2_2018_track1.py]: yes, agree 💯

**Review by galtay — APPROVED** [review]
thanks, nice work!
